neps/nep-0452.md

```diff
@@ -0,0 +1,220 @@
+
+The initial discussion can be found [here](https://gov.near.org/t/official-linkdrop-standard/32463/1).
+
+## Rationale and Alternatives
+
+* Why is this design the best in the space of possible designs?
+
+This design allows for flexibility and extensibility of the standard while providing a set of criteria that cover the majority of current linkdrop use cases. The design was heavily inspired by current, functional NEPs such as the Fungible Token and Non-Fungible Token standards.
+
+* What other designs have been considered and what is the rationale for not choosing them?
+
+A generic data struct that all drop types needed to inherit from. This struct contained a name and some metadata in the form of stringified JSON. This made it easily extensible for any new types down the road. The rationale for not choosing this design was both simplicity and flexibility. Having one data struct required keys to be of one type only when in reality, they can be many at once. In addition, having a generic, open-ended metadata field could lead to many interpretations and different designs. We chose to use a KeyInfo struct that can be easily extensible and can cover all use-cases by having optional vectors of different data types. The proposed standard is simple, supports drops with multiple assets, and is backwards compatible with all previous linkdrops, and can be extended very easily.
+
+A standard linkdrop creation interface. A standardized linkdrop creation interface would provide data models and functions to ensure linkdrops were created and stored in a specific format. The rationale for not choosing this design was that is was too restrictive. Standardizing linkdrop creation adds complexity and reduces flexibility by restricting linkdrop creators in the process in which linkdrops are created, and potentially limiting linkdrop functionality. The functionality of the linkdrop creation, such as refunding of assets, access keys, and batch creation, should be chosen by the linkdrop creator and live within the linkdrop creator platform. Further, linkdrop creation is often not displayed to end users and there is not an inherent value proposition for a standardized linkdrop creation interface from a client perspective. 
+
+* What is the impact of not doing this?
+
+The impact of not doing this is creating a fragmented ecosystem of linkdrops, increasing the friction for user onboarding. Linkdrop claim pages (e.g. wallet providers) would have to implement custom integrations for every linkdrop provider platform. Inherently this would lead to a bad user experience when new users are onboarding and interacting with linkdrops in general.
+
+## Specification
+
+```ts
+/// Information about a specific public key.
+type KeyInfo = {
+   /// How much Gas should be attached when the key is used to call `claim` or `create_account_and_claim`.
+   /// If not provided, a default value of 100 TGas will be attached to the transaction.
+   required_gas: string | null
```

Comment:
How do you see this gas being split between the calls? There are two arrays that require cross-contract calls (`NFTData[]` and `FTData[]`), and there is also some gas to be preallocated for the account creation callback and assets transfer callback functions, so there is a lot to unpack here.

P.S. It feels that it is quite trivial to end up with a linkdrop that won't have enough gas to transfer all the assets in a single transaction. Thus, one area for future possibilities would be a multi-stage withdrawal.

Reply:
> How gas is distributed should be left to the smart contract developer in my opinion.

@BenKurrek Well, this is exactly who I care about, the developers of the smart contract who would be in a tough situation where they only have the total gas allocation, and they need to somehow figure out how to spend it wisely. While alternatively, we might have `required_gas` specified on the FTData / NFTData level. How does Keypom handle it?

Reply:
> I strongly believe that the gas distribution and the inner workings of the logic should not be addressed in this standard.

After a second thought, I agree with this statement. Consider adding a hint to the readers that it is expected that it is the responsibility of a linkdrop contract to define how exactly to calculate the required gas (either automatically on the contract side with linkdrop creation, or even move this responsibility to the client code of linkdrop creation)